Refactor and add tests and metric to KillUnusedSegments duty (auto-kill) (#15941)
apache/druid: docs/operations/metrics.md

|`killTask/maxSlot/count`| Maximum number of task slots available for auto kill tasks in the auto kill run.                                                                                                                                                                                                                                                                                                                                                                                                                | |Varies|
|`kill/task/count`| Number of tasks issued in the auto kill run.                                                                                                                                                                                                                                                                                                                                                                                                                                                    | |Varies|

> |`kill/candidateUnusedSegments/count`|Number of candidate unused segments to be deleted from the metadata store in an auto kill run.|`dataSource`|Varies|

Reviewer: Reading at this doc and the name of the metric, it's not very clear what it represents.
Is it the number of segments that the coordinator _decided to kill_ in a datasource or the number of segments that _should be killed_ in a datasource? Also, should it have `interval` as a dimension?

Author: > Is it the number of segments that the coordinator decided to kill in a datasource or the number of segments that should be killed in a datasource? 

It's the latter - clarified the docs a wee bit. Let me know if you have a better suggestion. I called it candidate because it's ultimately the underlying kill task that can say this for sure.

>Also, should it have interval as a dimension?

No, there can be multiple intervals.

Reviewer: > No, there can be multiple intervals, so an interval dimension can explode the metric.

Makes sense.

|`killTask/maxSlot/count`| Maximum number of task slots available for auto kill tasks in the auto kill run.                                                                                                                                                                                                                                                                                                                                                                                                                | |Varies|
|`kill/task/count`| Number of tasks issued in the auto kill run.                                                                                                                                                                                                                                                                                                                                                                                                                                                    | |Varies|

> |`kill/candidateUnusedSegments/count`|The number of candidate unused segments eligible for deletion from the metadata store during an auto kill run for a datasource.|`dataSource`|Varies|

Reviewer: ```suggestion
|`kill/eligibleUnusedSegments/count`|The number of unused segments of a datasource that are identified as eligible for deletion from the metadata store by the coordinator.|`dataSource`|Varies|
```
